vocabularies/Hierarchy.xml

```diff
@@ -45,27 +45,136 @@
+            <String>The sibling rank of a node is the index of the node in the sequence of all nodes
+            in the unlimited hierarchy with the same parent. The first sibling has rank 0.</String>
+          </Annotation>
+        </Property>
+        <!--
+        <Property Name="SiblingOrderProperties" Type="Collection(Edm.PropertyPath)" Nullable="false">
+          <Annotation Term="Core.Description" String="Properties that define the relative order of siblings" />
+          <Annotation Term="Core.LongDescription" String="A PATCH request on these properties changes the relative order of siblings." />
+        </Property>
+        -->
+        <Property Name="MatchedProperty" Type="Edm.PropertyPath" Nullable="true">
+          <Annotation Term="Core.Description" String="Boolean property indicating [matching](#MatchCount) nodes" />
         </Property>
-        <Property Name="PreorderRankProperty" Type="Edm.PropertyPath" Nullable="true">
-          <Annotation Term="Core.Description" String="Property holding the preorder rank of a node" />
-          <Annotation Term="Core.LongDescription" String="The preorder rank of a node expresses its position in the sequence of nodes created from preorder traversal of the hierarchy structure after evaluating the $filter expression in the request excluding any conditions on key properties. The first node in preorder traversal has rank 0." />
+        <Property Name="MatchedDescendantCountProperty" Type="Edm.PropertyPath" Nullable="true">
+          <Annotation Term="Core.Description" String="Integer property of type `Edm.Int64` holding the the number of [matching](#MatchCount) descendants a node has in the unlimited hierarchy" />
         </Property>
+        <!--
+        <Property Name="StartNodeCountsSupported" Type="Core.Tag" Nullable="false" DefaultValue="false">
+          <Annotation Term="Core.Description" String="Whether the service provides the [`StartNodeCounts`](#StartNodeCounts) instance annotation" />
+        </Property>
+        -->
       </ComplexType>
 
+      <Term Name="MatchCount" Type="Edm.Int64" AppliesTo="Collection">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Common.IsInstanceAnnotation" />
+        <Annotation Term="Core.Description" String="Instance annotation on the result of an `$apply` query option containing the number of matching nodes after hierarchical transformations" />
+        <Annotation Term="Core.LongDescription">
+          <String>The service designates a subset of the `$apply` result as "matching nodes" derived from the start nodes
+          of `ancestors` and `descendants` as determined by the transformations in their fourth parameters.
+          This subset is typically the output set of a `filter` or `search` transformation that evaluates node
+          attributes other than the node value.&lt;br&gt;This instance annotation is available if [`RecursiveHierarchy/MatchedProperty`](#RecursiveHierarchyType)
```

Comment:
```suggestion
          This subset is the output set of the `filter` or `search` transformation in the fourth parameter of the last `ancestors` transformation.
          &lt;br&gt;This instance annotation is available if [`RecursiveHierarchy/MatchedProperty`](#RecursiveHierarchyType)
```
Would this be equivalent for our purposes? It is certainly easier to implement.

Reply:
Looks good to me. I agree that "ancestors" is the relevant transformation for this kind of search/filter: you need to locate matches and the complete the tree "upwards". A "descendants" transformation has a different semantics, it restricts to a sub-tree, but that is not what a "match" needs to indicate. You will notice that effect by looking at the resulting root node(s).